libraries/lts/include/mcrl2/lts/detail/liblts_bisim_m.h

```diff
@@ -0,0 +1,579 @@
+            neg_phi.negated = true;
+            set_truths(neg_phi);
+            return neg_phi;
+        } 
+
+        // Set of t derivates we need to take care of. 
+        std::set<block_index_type> dT; 
+        for (observation_t obs : blocks[b2].outgoing_observations) {
+            if (label(obs) == label(dist_obs)) {
+                dT.insert(target(obs));
+            }
+        }
+        std::vector<formula> conjunctions;
+
+        while (!dT.empty()) {
+            level_type max_intersect = 0; 
+            block_index_type splitBlock = *dT.begin();
+            for (block_index_type bid : dT ) {
+                if (gca_level(target(dist_obs), bid) > max_intersect) {
+                    splitBlock = bid;
+                    max_intersect = gca_level(target(dist_obs), bid);
+                }
+            }
+            formula f = distinguish(target(dist_obs), splitBlock);
+            conjunctions.push_back(f);
+
+            std::set<block_index_type> dTleft;
+            std::set_intersection(dT.begin(), dT.end(), f.truths.begin(), f.truths.end(),
+                std::inserter(dTleft, dTleft.begin())
+            );
+            assert(dT.size() > dTleft.size()); 
+            dT.swap(dTleft);
+        }
+        formula returnf;
+        returnf.conjunctions.swap(conjunctions);
+        returnf.label = label(dist_obs);
+        returnf.negated = false;
+        set_truths(returnf);
+        return returnf;
+    }
+
+    level_type gca_level(const block_index_type B1, const block_index_type B2) {
+        block_index_type b1 = B1, b2 = B2;
+        if (B1 < B2) {
+            b1 = B2;
+            b2 = B1;
+        }
+        std::pair<block_index_type, block_index_type> bpair(b1, b2);
+        if (greatest_common_ancestor.find(bpair) != greatest_common_ancestor.end()) {
+            return greatest_common_ancestor[bpair];
+        }
+        level_type lvl1 = blocks[b1].level, lvl2 = blocks[b2].level;
+        if (b1 == b2) {
+            greatest_common_ancestor.emplace(bpair, lvl1);
+            return lvl1;
+        }
+        block_index_type B1parent = b1, B2parent = b2;
+
+        if (lvl1 <= lvl2) {
+            B2parent = blocks[b2].parent_block_index;
+        }
+        if (lvl2 <= lvl1) {
+            B1parent = blocks[b1].parent_block_index;
+        }
+        if (B1parent == B2parent) {
+            lvl1 = blocks[b1].level;            
+        } else {
+            lvl1 = gca_level(B1parent, B2parent);
+        }
+        greatest_common_ancestor.emplace(bpair, lvl1);
+        return lvl1;
+    }
+
+    block_index_type lift_block(const block_index_type B1, level_type goal) {
+        block_index_type B = B1;
+        while (blocks[B].level > goal) {
+            B = blocks[B].parent_block_index;
+        }
+        return B;
+    }
+
+    /*
+    * \brief Converts the private formula data type to the proper mCRL2 state_formula objects
+    * \param[in] a formula f
```

Comment:
I think \param[in] confuses doxygen and this should just be \param (although this is being used in many other places as well). Also \retval should just be \return.